### ISSUE TITLE
"Quit RetroArch" is mapped to a key by default

### ISSUE THREAD

**Issue by RollingStar** [libretro/RetroArch#2670]
By default, RetroArch will immediately close down without saving anything if you press `Esc`. This perplexes me. I can't think of any game on Windows that has behavior like this. I've yet to encounter a time when the universal Windows combination `Alt + F4` does not work to close RetroArch. This key combination works in probably 99% of Windows software, so the user already knows it. Why map "quit" to one key? It's right next to the `F1` key which opens the menu (used all the time). Very easy for the end user to make a mistake.

**Comment by fr500** [comment]
This is not really a bug though, this is just a design decision, could be good or bad it's all a matter of perspective. I'll leave this open for a while to see if there is any feedback.

BTW, Alt + F4 works on my PC running windows 8.1 x64

**Comment by ProfessorKaos64** [comment]
Or just don't hit Esc. Map it in the menu to something else then.

**Comment by RollingStar** [comment]
@ProfessorKaos64, you're saying it's intended behavior that one of the most-used keys *in-game* (F1 = RetroArch menu) should be within 2 centimeters / 1 inch of the key that *instantly ends a user's game without saving*? Can you name another program that maps buttons with common functionality next to ones that end the program without saving? How many programs have a single key or button (not a combination) that closes everything without saving?

If there are no other such programs: what makes RetroArch so special? Is this a new user-interaction paradigm? Will Chrome, Microsoft Word, and Netflix copy this functionality within the next 5 years?

**Comment by joolswills** [comment]
takes 2 mins to remap it - less time than moaning on an issue tracker :-) Perhaps they will change the defaults at some point - or you could submit a pull request with your argument. Bearing in mind though that many people are used to the default now.

**Comment by twinaphex** [comment]
Escape to completely exit is a feature found in many Japanese PC games. FF13 PC on Steam has it like that, plenty of other games do too.

Not saying it was designed with that in mind but you can't go and argue it is not a convention of sorts when it is, it is just not something you are used to.

**Comment by fr500** [comment]
What makes RetroArch so special? 
So now we can do what we want in our own project?

I think there's no point debating this, it's just a default that you can change in a few seconds. Closed.

**Comment by twinaphex** [comment]
@Alcaro I told you several times I was open to adding this as a configurable option, don't be unreasonable now. I even suggested you could add it yourselves, but since I guess you don't want to do the work I will do it instead.

Honestly I'd appreciate some collaboration at times on RA. There are about 30 people in our main chan and only a handful that actually contributes anything beyond a few lines. Then how can you complain about this or that not being done? I'm not sure how much more time and effort I should put into this project to show you guys my passion and my commitment is sincere and I want to make it truly better. But if nobody else gives a shit then honestly what do you expect?

**Comment by ProfessorKaos64** [comment]
@twinaphex , wish I could help more, and hopefully in the future at some point. While I mainly script/code bash and create packages for SteamOS and am not a "traditional programmer", I highly appreciate the software. For what it's worth, thanks for sticking with the project despite being burned out.

**Comment by twinaphex** [comment]
I don't like also how this drama seeps over into our private channels and how teammembers I value just leave over the most insignificant of disagreements. I did not even insult anybody but there is so little regard for me or the time I make available for RA that they just leave over stupid stuff. 

I already indicated before I was going to add this. There was literally nothing to get 'pissed' over or to 'leave over'. People need to get a grip on themselves honestly and try walking a mile in my shoes once and see where it gets them. Then I can play quarterback referee and start judging everything you are doing. It's easy criticizing, it is harder leading by example.

**Comment by Alcaro** [comment]
You two close/dismiss an issue about dangerous keybindings with the motivations "it's standard in Japan" (I don't recall that being our primary target audience, and I doubt they put "quit without saving" right beside "open menu") and "you can change it" (read the title of the issue again). And then I'm the unreasonable one?

I don't mind slow or nonexistent progress. I do mind pushiness, dismissing user suggestions without reading them properly, dismissing everyone else's issues as minor no matter how the other party feels, your refusals to accept that humans are selfish creatures, and your obnoxious 'fuck that guy anyways' attitude whenever anyone dares leave.

You say I'm valued, but I don't feel that way. Whenever I'm relevant, it's usually because you're angry at someone and I'm trying to calm the situation down, or you're angry at me; never because I did something right. Is it really unreasonable to leave an environment of so much negativity?

If you want me around here, you need to respect that others don't always share your goals, and that this doesn't make them wrong or evil. If they don't want to do something (for example make a PR for this, or merge libretro into ScummVM), you can't force them; trying will only make everyone involved angry.

I believe that is the core of why you're involved in so much drama. I believe it is why you had to abandon the name Squarepusher. Fix it and I'll stay; don't, and I wish you best of luck finding more hard-headed contributors than me.

**Comment by fr500** [comment]
I close the issue because it has already de-railed and will most likely end up... well, like this. In my opinion it is not such a big deal to waste hours discussing.
In any project you'll see issues like this closed with a wont-fix tag.

Me, I would make ESC the menu toggle, but I don't want to impose my preference.

**Comment by twinaphex** [comment]
> You two close/dismiss an issue about dangerous keybindings with the motivations "it's standard in Japan" (I don't recall that being our primary target audience, and I doubt they put "quit without saving" right beside "open menu") and "you can change it" (read the title of the issue again). And then I'm the unreasonable one?

I just asked maister why he implemented 'escape = exit' and he mentioned that MPlayer does it like that and he preferred it that way.

That is the only reason really.

Also, you can rebind this option under Input Hotkey Settings. So again, I don't see what the actual problem here is.

> (read the title of the issue again)

"Quit RetroArch" is mapped to a key by default - that is the title. OK, so what is the big deal about that? Mplayer immediately exits when pressing Escape, just as us. MPV quits when pressing Q. Really, what is the issue here?

>  Is it really unreasonable to leave an environment of so much negativity?

You should have let me know of this then in PM. I do not get why you people always want to talk retroactively about how horrible your experience has been yet you never take the time to talk to me about it. I don't think it is fair to come at me with that when I am not even aware there is an issue. Talk to me about it one-on-one - I am never too arrogant or boastful not to take it into consideration or to change how I do things.

**Comment by Alcaro** [comment]
> Mplayer immediately exits when pressing Escape, just as us. MPV quits when pressing Q. Really, what is the issue here?

Closing your media player loses your position in the video, which you can easily find again. RetroArch contains complete games, and you can't go back and forth in games as in videos.

"It's configurable" is not a sufficient solution; new users don't check the config, they just fire it right up and expect it to work. If it doesn't, or if it deletes fifteen minutes of progress because their finger is misplaced, they deem it user hostile and go back to standalone emulators/etc. You can argue as much as you want that they should read the manual and check config, but that's how end users are; they're lazy, impatient bastards.

> I do not get why you people always want to talk retroactively about how horrible your experience has been yet you never take the time to talk to me about it.

Option 1: The thought didn't cross their mind. Humans act irrationally quite often.

Option 2: Ignoring or leaving is easier. Humans are lazy bastards. Every single one of them.

Option 3: They saw you as a hopeless case.

I believe all three of those apply, to varying degrees, to everyone who left. For me, it's mostly (1), with some amount of (2).

> I am never too arrogant or boastful not to take it into consideration or to change how I do things.
> If anything, my biggest fault is that I am too open and that I am too quick to merge PRs at times.

By not accepting my suggestion of what your worst flaw is, you have chosen to refuse my advice. From my viewpoint, that is arrogance. (From your viewpoint, I'm overreacting - which falls under "dismissing everyone else's issues as minor no matter how the other party feels".)

Bad code can be reverted. Lost contributors and friendships can not. Anything that leads to the latter is therefore a lot more important than the former.

> Out of interest, how many other projects do you know of where I give you guys so much power and control over projects? How many other projects do you know of that gives you co-ownership over an entire organization like here? And then you guys want to talk you are not being 'valued'. If you were not being valued, why the hell would I do that?

Access levels are a technical construction. How I feel treated is a social construction. You can't replace one with the other.

**Comment by twinaphex** [comment]
> By not accepting my suggestion of what your worst flaw is, you have chosen to refuse my advice. From my viewpoint, that is arrogance. (From your viewpoint, I'm overreacting - which falls under "dismissing everyone else's issues as minor no matter how the other party feels".)

Now I'm thinking you put words into my mouth - i never mentioned 'overreact'. 

Also, this is not the proper venue to have a serious discussion about this. You know where to contact me on IRC, I am very genuine about patching up whatever misgrievance there is, but I don't feel much for having this turn into some kind of public spectacle to be honest with you. If you don't even want to give me that opportunity then honestly it says you don't have much respect for me either after all this time and it starts feeling a lot like this whole 'make up gig' is entirely one sided.

**Comment by twinaphex** [comment]
To be perfectly honest with you Alcaro, since you didn't bother responding in PM on IRC, I'll tell you what I really feel about this stuff....

To be perfectly honest with you, I'm tired of all this stuff. I can either drop everything I am doing and just stop with RA, but that would be like taking my ball and going home. Or, this is the more preferable route, I just want to continue developing RetroArch, I want to continue developing libretro, and I don't want all this negativity around this project. I might have fuelled some of the negativity at some point or fed into it, but at this point I truly don't want it. It is a deterrent to what I want to accomplish, and if you know me truly well you can tell that I don't actually relish this atmosphere surrounding it.

Anyway, I cannot 'smell' if you guys are feeling not appreciated or you guys feel tainted by negativity. You should have brought them up to me when it happened, not after the fact. I cannot recall events from the past in full clarity that happened two or three years ago.

Do I have character flaws? Sure, plenty. So do you. So does everybody. Might I have done some bad shit in the past? Certainly. So did others. Does that make it right? No. If there is anything I have to apologize for, I will apologize for that right now. But I am done with arguing about this stuff anyway. I just want to focus on development and that is it. We are losing valuable time bickering over stuff of relatively little importance and no single enduser is helped by this stuff anyway.

I said before that I never wanted to 'trivialize' this user improvement, I do want to improve RA. I am just at times overwhelmed by the amount of work and I would appreciate some help. Is that so bad to ask for? You say there's an air of negativity surrounding the project, I could say the same. I see people on every forum and message board attacking us left and right, and I see an ever dwindling pool of contributors that even want to contribute, and then I have team members just walking out over basically nothing without even giving me a proper chance to resolve things or even find out where they are at emotionally. Sorry but that just isn't right and I want to get away from this negative crap.

So if you want to give this another chance in a negativity -free zone, you know where you can find me. Things were honestly going in the right direction the past few days so I don't know where this comes from all of a sudden. I don't want to talk about scene this scene that or character this character that, I want to get shit done. I want to work with simias to have Mednafen PSX use a GL renderer, I want a Dolphin libretro port done, I want RA 1.3.1 pushed out there, I want us to make the first tentative steps towards libretro v2. Basically, I want to get shit done and I'm going to get shit done. I might have done all things wrong in the past but talking about it and endlessly going back and forth over it is not going to help any, I can only move forward.

Anyway, I am just going to close this. I have made my intent clear, I do value you being onboard, if there was anything I said that rubbed you the wrong way I apologize, but honestly, I just want to get away from this negativity since it is helping nothing. You can give this another chance or you don't. Your choice really.

**Comment by twinaphex** [comment]
Removed a bunch of posts. To everyone - I have no more desire in any drama from this point on. Please keep it out of my sight OKthnx

**Comment by saviski** [comment]
How do I remap the "Quit RetroArch" HotKey to nothing? Can't find a way to set it to nothing through the UI.
I've managed to do it only by editing the input_exit_emulator key on the retroarch.cfg file to:
input_exit_emulator = "null"

**Comment by fr500** [comment]
press start on the setting you want to clear

**Comment by inukaze** [comment]
Hi there, i want remap "esc" instead "close retroarch (same if i press ALT+F4)" i prefer open the Quick Menu.

**Comment by fr500** [comment]
So... map it?, map quit to something else (or clear it with retropad Y) and map esc to menu toggle

**Comment by omenking** [comment]
Is there a user benefit to having the `esc` key mapped to quit by convention?
I ask genuinely as I would understand and accept its default mapping if there a reason other than user preference.

**Comment by Oggom** [comment]
> Is there a user benefit to having the esc key mapped to quit by convention?

Playing hentai games when you still live with your parents is a risky business.

**Comment by twinaphex** [comment]
Well, maister made the default bindings like that. I don't think maister had hentai games in mind when he did that, it just so happened to work the exact same way as what you would find in those games (also, games like FF13 have an 'Escape to immediately quit' convention too, so it's Japanese software in general) 

It has grown into a convention by now for retroArch, and if we change it now, dozens of downstream projects like RetroPie who probably designed around this would be in trouble.

Besides, I think the few Japanese users we recently gained are probably more familiar with this convention than others, and personally, I like the immediate exit thing.

Bottom line, you can change it if you don't like it. I don't see a big issue here.

**Comment by RollingStar** [comment]
What about saving state upon exit? How many cores support save states?

Or what about making the quick menu something other than F1? I just don't understand why two important functions with very different outcomes are next to each other.

Or what about asking for confirmation on exit? Or having a setting that does this?

**Comment by omenking** [comment]
What @twinaphex says makes sense. I wasn't sure the length and legacy of the keybindings.
As much as I'd like to see it changed, the damage to the long-term dependants might out weight the connivence to new comers.

I thought maybe for Windows/OSX specific builds its might be possible to serve up different default bindings.

**Comment by bparker06** [comment]
We tried making a "confirm on exit" option once, but due to the current state of the menu code, it caused all kinds of corner case failures, so we had to revert it.

**Comment by dishcandanty** [comment]
introducing myself to retro arch. the escape key closing the app is extremely annoying. a lifetime of applications has trained me to use escape to back out of a menu

**Comment by drone540** [comment]
dishcandanty same thing i was thinking. I think escape key would take me back a menu. i'm not sure what key going back to is mapped to besides delete. But how do you exit a current emulator and go back to the menu? is that even possible?

**Comment by omenking** [comment]
@dishcandanty @drone540 

https://github.com/libretro/RetroArch/issues/2670#issuecomment-322066748

**Comment by Ferk** [comment]
Doesn't retroarch.cfg get automatically created for the user when it's freshly installed?
New defaults would only affect new installs, nobody's configuration would get changed.
I understand that some people might be surprised when they reinstall their retroarch, though.

Downstream projects like RetroPie should provide their own default if they really depend in configurations in such a way that the system actually is "designed around it", otherwise it wouldn't be a very robust design, imho. But considering this setting can be changed easily it wouldn't be a very sensible design decision to depend on that, imho.

@drone540 It's F1 to go back to the menu.

**Comment by bparker06** [comment]
@RollingStar We already have an 'auto-save state' that works on exit I believe, it should be in Settings -> Saving.

**Comment by drone540** [comment]
@Ferk Yeah it was F1 I was looking for.
@RollingStar Yeah F1 and Escape are pretty close together. I can see easily hitting the wrong one by accident.
@bparker06 Does that auto-resume when you start the game back up? Or how would you use that auto-save state.

I guess the Escape issue is less likely when you are using a controller to play the game. I installed retro arch at my job on linux before my home, so playing with the keyboard as controller just makes me want to hit escape to do any menu actions since it's more natural when you are already on the keyboard. But guess I'll have to be more careful and save my games more often before messing with stuff.

**Comment by BenMcLean** [comment]
This is absolutely crazy. On TyrQuake, you need the ESC key to access the menu.

**Comment by twinaphex** [comment]
You can use Game Focus toggle (F11) in order to avoid this.

**Comment by dankcushions** [comment]
my era of PC games (90s, early 00s) had the convention of Escape being the 'menu' key so this default has often caught me out, even though i'm aware of it. i'm not sure of what the convention is these days, but i would guess escape still means 'menu' in most PC games, and you exit game through the relevant menu option.

i know you can change it, but escape = immediate exit does seem a pretty niche default. it's one of those little things that often comes up in retroarch complaints.

**Comment by edenprojectde** [comment]
This is an insane design decision, i just downloaded it and i cant get anything to  work because i always have in mind that ESC opens the Menu or goes back a menu entry. This is true madness. I hate this.

**Comment by inukaze** [comment]
Mee too, i ever press ESC key to Open the Quick Menu instead F1. the Exit combination should be another like F9 or F12

**Comment by fr500** [comment]
Just change it....

**Comment by inukaze** [comment]
i never change the default mappings xD of RetroArch GUI. because if i need to install and/or use the software in another device or pc. the most people don't change neither. 

For example, the mapping of Mupen64Plus Core, in Mario Kart 64, with my Xbox360 Gamepad, when i press "RT" have 2 functions, "Z+B" (Use item, and brake), and yes, at same time. xD

i don't know why the default Mapping is it xD. but the Gamepad Mapping, yes i change for some games and/or specific cores.

**Comment by BenMcLean** [comment]
There ought to be some key combo to instantly quit but it needs to be something which isn't used by the cores.

**Comment by bparker06** [comment]
I think this thread has run its course and there is nothing left to discuss.

The default is not going to be changed but there are currently several ways to change this behavior at runtime for those that don't like it:

- remove the Esc bind entirely
- change the bind to a different key you like better
- use game focus toggle
- use hotkey enable

For everyone that disagrees there are as just many that prefer it like it is. I think a default with so many ways to change the behavior is already more than enough. I'm locking this thread.
